2-truncated quasi-categories

juliankom opened the pull request:
This PR defines a new structure `Quasicategory₂` which follows closely Definition 1.3.1 of the [blueprint](https://emilyriehl.github.io/infinity-cosmos/blueprint/sec-homotopy-category.html). A proof of Lemma 1.3.2 is given, namely that the 2-truncation of a `Quasicategory` is a `Quasicategory₂`. 

The (3, 1)- and (3, 2)-filling properties get their own separate structures (`horn3[1/2].horn_data` and `horn3[1/2].filling_simplex`) to be able to more elegantly state and prove lemmas relating these properties to filling 3-horns. This is done within the `TwoTruncated.lean` file.

The proof is complete up to the following: To use the definition of a `Quasicategory`, one needs to construct maps from inner horns given the combinatorial data of `Quasicategory₂`. In this PR, I do this using universal properties of 2- and 3-horns, but I do not prove the universal properties themselves, as this is being worked on by Joël Riou, see e.g. [this PR](https://github.com/leanprover-community/mathlib4/pull/23872). This means that in the future, the `sorry`s concerning these universal properties should be filled in once that PR goes through. The construction of these diagrams is done in the `Horn.lean` file.

Moreover, I am opening this PR as an early draft for comments, since this is my first time contributing to this library. There are still various TODOs that I am still working on, and certainly some definitions and lemmas will need to be put into appropriate files and renamed.

Closes #103

emilyriehl commented:
We're not as picky as mathlib but to streamline eventual PRs into mathlib, it's probably good to follow their style guides regarding terming naming: https://leanprover-community.github.io/contribute/naming.html

In particular, I believe your structures like `horn_data` should be in upper camel case as `HornData`.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,394 @@
+open horn₂₁
+namespace horn₂₁
+
+def path_edge₀ {X : SSet} (f : Path X 2) : Δ[1] ⟶ X := yonedaEquiv.symm (f.arrow 1)
```
Defs like this should be lower camel case. Since the input data is a path rather than a 2-simplex something like `pathEdge₀ f := yonedaEquiv.symm (f.arrow 0)` feels more intuitive to me but feel free to disagree.

emilyriehl replied:
Also if you turn the `def` into an `abbrev` then you can cut the dsimps on line 139.

juliankom replied:
> Defs like this should be lower camel case. Since the input data is a path rather than a 2-simplex something like `pathEdge₀ f := yonedaEquiv.symm (f.arrow 0)` feels more intuitive to me but feel free to disagree.

I agree that this is numbering is more intuitive. I will make sure that usage is consistent.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,394 @@
+def path_edge₂ {X : SSet} (f : Path X 2) : Δ[1] ⟶ X := yonedaEquiv.symm (f.arrow 0)
+
+def path_edges_comm {X : SSet} {f : SSet.Path X 2} : pt₁ ≫ path_edge₀ f = pt₀ ≫ path_edge₂ f := by
+    dsimp only [pt₀, pt₁, path_edge₀, path_edge₂]
```
More trivial style comments: these three lines should only be indented once. (Once for proofs, twice for theorem/def statements.)

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,394 @@
+the inclusion Λ[3, 1] → Δ[3], there exists a 2-simplex satisfying the (3, 1)-filling property
+(namely, `yonedaEquiv g`).
+-/
+def fill31_from_horn_extension (g : Δ[3] ⟶ X) (h : horn_from_data horn_data = Λ[3, 1].ι ≫ g)
```
I think this formatting for the proof is more readable (and likely could be improved further):
```
def fill31_from_horn_extension (g : Δ[3] ⟶ X) (h : horn_from_data horn_data = Λ[3, 1].ι ≫ g)
  : ∃ σ : ((truncation 2).obj X) _⦋2⦌₂, Truncated.fill31.filling_simplex horn_data σ := by 
  let σ := X.map (SimplexCategory.δ 1).op (yonedaEquiv g)
  use σ
  exact {
    edge₀ := by 
      have arr : f.arrow 2 = (f.interval 1 2).arrow 1 := rfl
      rw [arr, ← horn_data.h₀, Truncated.spine_arrow, mkOfSucc_2_1]
      dsimp only [truncation, SimplicialObject.truncation, inclusion, whiskeringLeft_obj_obj, len_mk,
        id_eq, Functor.comp_obj, Functor.op_obj, fullSubcategoryInclusion.obj, Nat.reduceAdd,
        Fin.isValue, tr, Functor.comp_map, Functor.op_map, Quiver.Hom.unop_op,
        fullSubcategoryInclusion.map]
      have : yonedaEquiv.symm horn_data.σ₀ = stdSimplex.δ 0 ≫ g
          := by rw [← mcofork_up0 horn_data, h, ← Category.assoc, horn₃₁.incl₀]
      rw [← FunctorToTypes.map_comp_apply, ← op_comp, push_yonedaEquiv _ horn_data.σ₀ this]
      rfl
    edge₁ := by
      dsimp only [truncation, SimplicialObject.truncation, inclusion, whiskeringLeft_obj_obj, len_mk,
      id_eq, Functor.comp_obj, Functor.op_obj, fullSubcategoryInclusion.obj, Nat.reduceAdd,
      Fin.isValue, tr, Functor.comp_map, Functor.op_map, Quiver.Hom.unop_op,
      fullSubcategoryInclusion.map]
      have : yonedaEquiv.symm horn_data.σ₂ = stdSimplex.δ 2 ≫ g
          := by rw [← mcofork_up2 horn_data, h, ← Category.assoc, horn₃₁.incl₂]
      rw [← FunctorToTypes.map_comp_apply, ← op_comp, push_yonedaEquiv _ horn_data.σ₂ this]
      apply congr_fun
      apply Prefunctor.congr_map
      apply (Opposite.op_inj_iff _ _).2
      symm; exact @SimplexCategory.δ_comp_δ 1 1 1 (by norm_num)
    edge₂ := by 
      dsimp only [truncation, SimplicialObject.truncation, inclusion, whiskeringLeft_obj_obj, len_mk,
      id_eq, Functor.comp_obj, Functor.op_obj, fullSubcategoryInclusion.obj, Nat.reduceAdd,
      Fin.isValue, tr, Functor.comp_map, Functor.op_map, Quiver.Hom.unop_op,
      fullSubcategoryInclusion.map]
      have : yonedaEquiv.symm horn_data.σ₃ = stdSimplex.δ 3 ≫ g
          := by rw [← mcofork_up3 horn_data, h, ← Category.assoc, horn₃₁.incl₃]
      rw [← FunctorToTypes.map_comp_apply, ← op_comp, push_yonedaEquiv _ horn_data.σ₃ this]
      apply congr_fun
      apply Prefunctor.congr_map
      apply (Opposite.op_inj_iff _ _).2
      symm; exact @SimplexCategory.δ_comp_δ 1 1 2 (by apply Fin.le_iff_val_le_val.2; norm_num)
  }  
```

juliankom commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,394 @@
+-- and the proofs could probably be greatly simplified
+section aux_lemmas
+
+lemma map_yonedaEquiv {n m : ℕ} {X : SSet} (f : .mk n ⟶ .mk m) (g : Δ[m] ⟶ X) : X.map f.op (yonedaEquiv g)
```
Thanks, I have checked out the mathlib linebreak conventions and will update everywhere.

juliankom commented:
I have now ported everything to the `CompStruct` formalism and documented most (but not all) relevant definitions and lemmas. Modulo moving/renaming some stuff and adhering to code standards (although I hope I did an okay job at this), this should now be complete.

juliankom commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,394 @@
+
+end Truncated
+
+-- TODO: this section contains 3 lemmas moving application of yonedaEquiv around.
```
Because I don't want to deal with a mathlib PR right now, I moved this stuff into `StdSimplex.lean`. I (or someone else) might try to PR into mathlib at some point in the future.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+
+namespace SSet
+
+def horn_face_incl {n : ℕ} (i : Fin (n + 2)) (j : Fin (n + 2)) (h : j ≠ i)
```
`hornFaceIncl`

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+-/
+namespace horn₂₁
+
+def ι₀ : Δ[1] ⟶ Λ[2, 1] := horn_face_incl 1 0 (by norm_num)
```
Maybe better as abbreviations?

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+  symm
+  exact @stdSimplex.δ_comp_δ _ _ _ 0 1 (by norm_num)
+
+def horn_pushout : Limits.PushoutCocone pt₁ pt₀
```
`pushout` (since you're in the horn namespace)

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+  := Limits.PushoutCocone.mk ι₀ ι₂ sq_commutes
+
+-- TODO Joel's PRs (see e.g. mathlib PR #23872) proves this and more
+def horn_is_pushout : Limits.IsColimit horn_pushout := by sorry
```
`pushoutIsPushout`

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+
+open SimplexCategory
+
+def multispan_index : Limits.MultispanIndex J SSet where
```
All these defs should be lowerCamelCase.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/SimplicialSet/Horn.lean:
```diff
@@ -0,0 +1,158 @@
+/--
+Index types for the left and right coproducts in the coequalizer.
+-/
+def R := { x : Fin 4 // x ≠ 1 }
```
I'm not sure what the best conventions are for local names like this. Let's not stress now but perhaps revisit later.

juliankom replied:
Since the whole `Horn` file will have to be eventually revisited/reworked to complete the pushout and coequalizer sorries, I would suggest to leave it as is for now.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,433 @@
+variable {S : SSet} {x₀ x₁ x₂ : ((truncation 2).obj S) _⦋0⦌₂}
+  (e₀₁ : Edge x₀ x₁) (e₁₂ : Edge x₁ x₂)
+
+def path_edges_comm : pt₁ ≫ edge_map e₁₂ = pt₀ ≫ edge_map e₀₁ := by
```
Should be a theorem or lemma not a def.

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,433 @@
+to the two edges (this is made precise in the lemmas `horn_from_edges_restr₀` and
+`horn_from_edges_restr₁`).
+-/
+def horn_from_edges : Λ[2, 1].toSSet ⟶ S :=
```
`fromEdges` (since you're in the horn21 namespace)

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,433 @@
+and edge `e₀₂` such that `e₀₁`, `e₁₂`, `e₀₂` bound a 2-simplex of `S` (this is witnessed
+by `CompStruct e₀₁ e₁₂ e₀₂`).
+-/
+def fill21_from_horn_extension
```
lowerCamelCase

emilyriehl commented on InfinityCosmos/ForMathlib/AlgebraicTopology/Quasicategory/TwoTruncated.lean:
```diff
@@ -0,0 +1,433 @@
+open horn₃₁ (fill31_from_horn_extension)
+open horn₃₂ (fill32_from_horn_extension)
+
+-- TODO should this be SSet namespace or SSet.Truncated?
```
I think the `SSet.Truncated` namespace.

juliankom commented:
I updated the naming as suggested. Also, I split off some of the two lemmas about objects of `Truncated m` into the `SimplicialSet.Basic` file, where I believe they belong.

juliankom commented:
Note that I bumped mathlib while working on this project. Maybe that should be a commit on the `main `branch before merging this.

juliankom commented:
This PR now matches `main` mathlib version and builds.
